dlib/function.h

```diff
@@ -0,0 +1,91 @@
+// Copyright (C) 2022  Davis E. King (davis@dlib.net)
+// License: Boost Software License   See LICENSE.txt for the full license.
+#ifndef DLIB_FUNCTION_H_
+#define DLIB_FUNCTION_H_
+
+#include <cstddef>
+#include "te.h"
+#include "invoke.h"
+
+namespace dlib
+{
+    template <
+        class Storage, 
+        class F
+    > 
+    class function_basic;
+
+    template <
+        class Storage,
+        class R, 
+        class... Args
+    > 
+    class function_basic<Storage, R(Args...)> 
+    {
+    public:
+        constexpr function_basic(std::nullptr_t) {}
+        constexpr function_basic()                                          = default;
+        constexpr function_basic(const function_basic& other)               = default;
+        constexpr function_basic& operator=(const function_basic& other)    = default;
+        constexpr function_basic(function_basic&& other)                    = default;        
+        constexpr function_basic& operator=(function_basic&& other)         = default;
+
+        template <
+            typename F,
+            std::enable_if_t<!std::is_same<std::decay_t<F>, function_basic>::value &&
+                             dlib::is_invocable_r<R, F&&, Args...>::value,
+                             bool> = true
+        >
+        function_basic(
+            F&& f
+        ) : str{std::forward<F>(f)},
+            func{[](void* self, Args... args) -> R {
+                return dlib::invoke(*reinterpret_cast<std::add_pointer_t<F>>(self),
+                                    std::forward<Args>(args)...);
+            }}
+        {
+        }
+
+        template <
+            typename F,
+            std::enable_if_t<!std::is_same<std::decay_t<F>, function_basic>::value &&
+                             dlib::is_invocable_r<R, F&&, Args...>::value,
+                             bool> = true
+        >
+        function_basic& operator=(F&& f)
+        {
+            *this = std::move(function_basic{std::forward<F>(f)});
+            return *this;
+        }
+
+        explicit operator bool() const noexcept
+        {
+            return !str.is_empty() && func;
+        }
+
+        R operator()(Args... args) const {
+            return func.ptr(const_cast<void*>(str.get_ptr()), std::forward<Args>(args)...);
+        }
+
+    private:
+        Storage str;
+        te::managed_move_pointer<R(void*, Args...), true> func;
```

Comment:
```suggestion
        R (*func)(void*, Args...) func = nullptr;
```
Why not this?  Like why have `managed_move_pointer`?

Reply:
So that a function is always empty after move even if using stack based memory. I wasn't sure about that. I'm tempted to remove it.

Reply:
The idea was to have a simple type that managed what happened to the pointer after a move operation.
I thought it would make people's lives easier to control the state of their type-erased objects after move. For example a function like `is_empty()` could be defined as:
```
bool is_empty() const { return storage.is_empty(); }
``` 
But then, depending on the storage type, `is_empty()` might have different values after move. Indeed, using `storage_heap`, then after move, `is_empty() == false`, but if using `storage_stack`, then `is_empty() == true` after move (if the original object contained something).

So i thought, maybe if your custom function pointers were managed by `managed_move_pointer` and `is_empty()` was redefined as:
```
bool is_empty() const { return storage.is_empty() && func; }
```

then after a move operation, `is_empty() == false` always, regardless of storage type. It just ensures a bit more consistency across storage types.

But, i'm not sure where I stand here. In my opinion, if an object is valid, then it's valid, and you shouldn't make it artificially not valid. But that means you can't guarantee what the state after move will be, other than it's "valid". But that's precisely what move is supposed to be: "valid" but "undetermined".
So maybe we never need `managed_move_pointer`.

Then the idea of having `null_after_move=false` was to easily switch what moved did. If `null_after_move=false` then it's just the same as vanilla pointer.

So there you go, you get to configure what the state after move is a bit more.

Opinions or ideas?

Reply:
That's awfully subtle though.  And it doesn't look like the code in here needs that.  Like the move assignment operator in `storage_heap` uses `std::exchange()` to set the source to `nullptr` so it's all good already.  And that's better since it's real clear from looking at that assignment operator that it is correct.  I.e. it doesn't require reasoning about `null_after_move` or any of that other stuff to figure out if it's really working. 

But yeah, it looks like `storage_stack::is_empty()` always returns false right now.  It should definitely say false when it doesn't have anything in it. Like the meaning of `is_empty() == false` is "you can get the thing out of it or call the contained function".  Which is definitely not the case for a default initialized object that's some kind of std::function equivalent.

State after move can be any valid thing so long as the interface contract clearly says what it is.  I would go with being empty though.